Add support for yml filenames where locale is used as prefix

1st8 opened the pull request.

Commit:
Add support for yml files with format locale.something.yml
Extract path transformation helper #replace_locale_in_path

1st8 commented:
Any thoughts about this? I don't want to annoy anyone, but please don't leave a small and focused PR hanging like this :wink:

kaspth commented on lib/i18n_yaml_editor/store.rb:
```diff
@@ -76,14 +76,10 @@ def create_missing_keys
+          # this just replaces the locale part of the file name
+          path = translation.file
+              .sub(/(\/|\.)#{translation.locale}\.yml$/, "\\1#{locale}.yml")
+              .sub(/\/#{translation.locale}([^\/]+)\.yml$/, "/#{locale}\\1.yml")
```
Find these regexes really tough to read, when what we're trying to do is remove the `-2` constant. How about:

``` ruby
parts = File.basename(translation.file).split('.')
parts[parts.index(translation.locale)] = locale

path = "#{File.dirname(translation.file)/#{parts.join('.')}"
```

:grin:

kaspth commented:
@1st8 Ha, had that coming :smile:

1st8 commented:
@kaspth Hehe, yeah :wink: 
Your suggested code is better, indeed. Changed the implementation to use it and extracted it out into a method to make it more readable. Updated the specs as well.

kaspth commented on test/unit/test_store.rb:
```diff
@@ -102,6 +102,22 @@ def test_create_missing_translations_in_top_level_file
+    @store.create_missing_keys
+
+    assert_equal(1, calls.count)
+    assert_equal(['da', 'en', '/tmp/da.yml'], calls.first)
```
Don't see the point in testing argument positions, sounds like a bag of hurt going forward :grin:

1st8 replied:
I am testing, that the interface defined by `Transformation. replace_locale_in_path` is used and used correctly. Otherwise I can't be sure, that it is actually used and not being replaced with something inline, which would make my encapsulated tests obsolete.

kaspth commented on test/unit/test_store.rb:
```diff
@@ -102,6 +102,22 @@ def test_create_missing_translations_in_top_level_file
+    calls = []
+    @store.define_singleton_method(:replace_locale_in_path) do |*args|
+      calls << args
+      "/tmp/special_path/en.yml"
```
You're just testing that you've stubbed a method. No go. Can't we just call the method, and assert the outputted path? :smile:

1st8 replied:
Yeah, see the two tests at `unit/test_transformation.rb:45` for that.

kaspth commented on lib/i18n_yaml_editor/transformation.rb:
```diff
@@ -37,5 +37,25 @@ def nest_hash hash
+      parts = File.basename(path).split('.')
+      parts[parts.index(from_locale)] = to_locale
+      File.join(File.dirname(path), parts.join('.'))
+    end
```
Method extraction is fine, but what's the use case for making this public API? Would rather it was just a private method in `store.rb`.

kaspth replied:
Also think `path` comes more naturally as the first argument.

1st8 replied:
I like to use IYE not only through the browser frontend, but in migration scripts etc. too. So having the `nest_hash` and `flatten_hash` methods available was pretty useful and the same applies to this new method in my opinion.
Besides, why not make it public? It is documented and tested.

kaspth commented:
Coming along nicely :+1:
